### ISSUE TITLE
[❔ other question] Caustic design issues

### ISSUE THREAD

**Issue by SouthEndMusic** [mitsuba-renderer/mitsuba2#580]
Hi,

I'm attempting to create a caustic optimization loop as in the tutorial, with the key difference that my lens surface is defined by parameters that are generated by a neural network. The implementation of the network and the construction of the surface mesh are in PyTorch.
At the moment I have some difficulties with setting up the scene:
- Is there a way to create the directional area light from the tutorial with the current Mitsuba 2 version?
- Is there a way to introduce the lens surface mesh from PyTorch directly into the scene? Now I create a `mitsuba.render.Mesh` object which writes the object to a ply file, and then load it from disk together with a scene description using `mitsuba.core.xml.load_dict`. I want my whole pipeline to be differentiable for optimization.
- I would also like to use `ptracer` as an integrator as in the tutorial, any idea when that will be available for Mitsuba 2?

System information:
- Windows 11
- mitsuba variant: gpu_autodiff_mono

Thanks in advance!

**Comment by merlinND** [comment]
Hi @SouthEndMusic,

Sorry that the current public version is missing some important pieces. Both the directional area light and the `ptracer` will be part of the upcoming release (#490). Then the tutorial you are referring to should run out of the box.

In order to get e.g. mesh vertices from PyTorch while remaining end-to-end differentiable, you will have to make sure that the quantities output by PyTorch are passed directly to the Mitsuba mesh instance. Writing them to disc and reloading them will break the computation graph.
Calling `traverse(mesh)` should return a map of parameters that can be updated directly in the mesh. For vertex positions directly, I'm not sure how well that works currently although it's supported better in the upcoming version as well.

**Comment by SouthEndMusic** [comment]
Thanks for your reply. I have now switched to a point source but it doesn't seem that any light from it passes through the lens (with the `path` integrator). Do you have any idea why? I did manage to make a render of the lens itself with an ambient emitter.
![Lens_with_scattering](https://user-images.githubusercontent.com/74617371/159002041-cfadc12c-9fef-4ba9-9478-bd1fbf0ef98b.png)

**Comment by Speierers** [comment]
Hi @SouthEndMusic,

If you material is a pure dielectric, then the `path` integrator won't be able to connect a shading point on the surface of this object to the point light (e.g. integral of the product of two delta dirac functions). For this to work you will need to use a particle tracer. In the meantime you could also use a finite light source, like a small `area` emitter.

**Comment by SouthEndMusic** [comment]
Hi @Speierers,

Thanks for your reply. With a small `area` emitter and the `path` integrator, I indeed get light through the lens, but I don't get the caustics I expect:

![f36f9911-ab58-4435-8595-7c5a3fc328a1](https://user-images.githubusercontent.com/74617371/159433499-c0caf986-132a-4df2-bc11-9909943638f6.jpg)

The above was made with just one flat lens surface. Is this maybe a sampling issue?

**Comment by Speierers** [comment]
Depending how small your area light is, you'll need a lot of samples. Try to increase the number of spp or the size of the area light

**Comment by SouthEndMusic** [comment]
@Speierers thanks, increasing the spp gives a more uniform illumination. Now I have some new issues:
- The rendering takes very long, possibly because I was only able to compile Mitsuba in debug mode
- Now I'm trying to render with `mitsuba.python.autodiff.render`, I get an error pop-up entitled 'Microsoft Visual C++ Runtime Library' with the following message:

> Assertion failed!
> 
> Program: path\to\mitsuba2\dist\mitsuba-render.dll
> File: path\to\mitsuba2\ext\enoki\include\enoki/array_idiv.h
> Line: 161
>
> Expression: d!= 1

**Comment by Speierers** [comment]
> The rendering takes very long, possibly because I was only able to build Mitsuba 2 in debug configuration

Yes probably

> Now I'm trying to render with mitsuba.python.autodiff.render, I get an error pop-up entitled 'Microsoft Visual C++ Runtime Library' with the following message:

I will need more information to help you with this. Could you send me the whole log please?

**Comment by SouthEndMusic** [comment]
I put the error for the 'release' build of Mitsuba in the previous post. I'm haven't been able to produce the log you requested, could you explain how I can produce this?

**Comment by Speierers** [comment]
I am simply interested to see exaclty where the error message occurs in the Python code. Do you have a stack trace maybe?

**Comment by SouthEndMusic** [comment]
@Speierers I'm running my code from a Jupyter notebook. When trying to do run a differentiable rendering I get the above pop-up, with the choices Abort/Retry/Ignore. Abort/Retry leads to a dead kernel, Ignore leads to an empty image. At no point do I get any traceback.

Even when I run the code from the command line with extra verbosity do I not get more information, just the usual

`cuda_eval(): launching kernel (n=4, in=4, out=3, ops=31)
cuda_eval(): launching kernel (n=1, in=0, out=2, ops=2)
cuda_eval(): launching kernel (n=1875, in=2, out=2, ops=18)
cuda_eval(): launching kernel (n=150, in=8, out=8, ops=72)
cuda_eval(): launching kernel (n=100, in=4, out=4, ops=36)
cuda_eval(): launching kernel (n=12, in=1, out=1, ops=9)
cuda_eval(): launching kernel (n=48, in=5, out=24, ops=358)
cuda_eval(): launching kernel (n=50, in=4, out=3, ops=31)
cuda_eval(): launching kernel (n=48, in=5, out=24, ops=358)
cuda_eval(): launching kernel (n=50, in=4, out=3, ops=31)
cuda_eval(): launching kernel (n=48, in=5, out=24, ops=358)
cuda_eval(): launching kernel (n=50, in=4, out=3, ops=31)
cuda_eval(): launching kernel (n=48, in=5, out=24, ops=358)`

until the error appears. The error also always contains the following information:

> For information on how your program can cause an assertion 
> failure, see the Visual C++ documentation asserts
>
> Press Retry to debug the application - JIT must be enabled

Do you by the way have any idea how to fix the release instead of debug build problem? Thanks in advance.

**Comment by Speierers** [comment]
`mitsuba.python.autodiff.render` is implemented in Python in the [this file](https://github.com/mitsuba-renderer/mitsuba2/blob/master/src/python/python/autodiff.py). You could for instance run your script in a Python debugger to better understand where the problem is occuring. Or using the good old `print` debugging technique 😉

**Comment by SouthEndMusic** [comment]
@Speierers with the debugger in Spyder I have been able to trace the problem to line 25 of `mitsuba2\dist\python\mitsuba\python\autdiff.py`:

`sampler.seed(0, total_sample_count)`

In [the docs](https://mitsuba2.readthedocs.io/en/latest/generated/render_api.html?highlight=sampler%20seed#mitsuba.render.Sampler.seed) it says something about a condition on `wavefront_size` which I do not understand, does it have something to do with `samples_per_pass` in the integrator?

**Comment by Speierers** [comment]
`samples_per_pass` should only be necessary when you are using such a high number of samples per pixel that it saturated your GPU memory. In which case it will split the render in multiple passes. At what resolution are you rendering, and with how many sample per pixels? Also what's your system specs?

**Comment by SouthEndMusic** [comment]
I only considered `samples_per_pass` since it is used in the caustics optimization tutorial (removing the value does not solve the differentiable rendering problem). I render not much larger than 100x100 since I ran into Cuda memory allocation errors quite quickly, even before I cranked up the spp to get a decent amount of light through my lens. I have [this](https://www.techpowerup.com/gpu-specs/quadro-t1000-max-q.c3807) GPU and [this](https://ark.intel.com/content/www/us/en/ark/products/201837/intel-core-i710750h-processor-12m-cache-up-to-5-00-ghz.html) CPU.

**Comment by SouthEndMusic** [comment]
I made some progress: the differentiable rendering issue was fixed by switching from the stratified to the independent sampler. I also switched to `mitsuba.python.autodiff.render_torch`. The existance of a `grad_fn` in the resulting image tensor indicates some succes:
`tensor([<data>], device='cuda:0', grad_fn=<ReshapeAliasBackward0>)`

I pass the scene parameters which respect to which I want to differentiate to `render_torch` as discussed in 'Integration with PyTorch' in the docs.  Is that also what I should do if I want to propagate gradients all the way back to a neural network? Now if I try backpropagation I get this error:

> render_torch(): critical exception during backward pass: set_gradient(): no gradients are associated with this variable (a prior call to requires_gradient() is required.) 

The differentiability breaks at the rendering, since I am able to train my neural network if I replace the rendering and the subsequent loss function with a trivial loss function. `requires_grad` is only set to true for the parameters of the neural network.  I set the data for the lens surface mesh objects like this:

```python
params = traverse(self.scene)
params["lens_exit.vertex_positions_buf"][:] = mi.core.Float32(self.lens.vertices["exit"].reshape(-1))
...
params.update()
```

here `self.lens.vertices["exit"]` has an intact computational graph back to the neural network.

In earlier discussions I found references to an integrator called `pathreparam` for geometric optimisation but I couldn't find anything on that in the docs, is that only for a previous Mitsuba 2 release?

**Comment by SouthEndMusic** [comment]
I'm trying to implement something analogous to the example in the link above, does this make any sense? `S` is the object in which I construct my Mitsuba scene.

 ```python
class diffrender(torch.autograd.Function):
    
    @staticmethod
    def forward(ctx, S):
        
        params = traverse(S.scene)
                
        ctx.in_exit_vertex_positions = params['lens_exit.vertex_positions_buf']
        ctx.in_exit_vertex_normals   = params['lens_exit.vertex_normals_buf']
        
        # Get the parameters with respect to which must be differentiated
        params_deriv = ['lens_exit.vertex_positions_buf',
                        'lens_exit.vertex_normals_buf']
        
        if S.lens.closed_mesh:
            for name in ["neg_x", "pos_x", "neg_y", "pos_y"]:
                params_deriv += [f"lens_{name}.vertex_positions_buf"]
                
            ctx.in_neg_x = params['lens_neg_x.vertex_positions_buf']
            ctx.in_pos_x = params['lens_pos_x.vertex_positions_buf']
            ctx.in_neg_y = params['lens_neg_y.vertex_positions_buf']
            ctx.in_pos_y = params['lens_pos_y.vertex_positions_buf']
        
        # params.keep(params_deriv)
        # params_torch = params.torch()
        
        arrays = [ctx.in_exit_vertex_positions,
                  ctx.in_exit_vertex_normals]
        
        if S.lens.closed_mesh:
            arrays += [ctx.in_neg_x, ctx.in_pos_x,
                       ctx.in_neg_y, ctx.in_pos_y]
        
        for param in arrays:
            ek.set_requires_gradient(param, True)
        
        ctx.out = render(S.scene, spp = S.sample_count)#, params = params, **params_torch)
        
        out_torch = ctx.out.torch()
        
        ctx.closed_mesh = S.lens.closed_mesh
        
        return out_torch
    
    @staticmethod
    def backward(ctx, grad_out):
        
        ek.set_gradient(ctx.out, ek.FloatC(grad_out))
        
        ek.cuda_autodiff.FloatD.backward()
        
        result = [ek.gradient(ctx.in_exit_vertex_positions),
                  ek.gradient(ctx.in_exit_vertex_normals)]
        
        del ctx.in_exit_vertex_positions, ctx.in_exit_vertex_normals
        
        if ctx.closed_mesh:
            for array in [ctx.in_neg_x, ctx.in_pos_x, ctx.in_neg_y,ctx.in_pos_y]:
                result.append(ek.gradient(array))
                
            del ctx.in_neg_x, ctx.in_pos_x, ctx.in_neg_y, ctx.in_pos_y
                
        return tuple(result)
```
